### DISCUSSION TITLE
Working group proposal: OIDC improvements

### DISCUSSION THREAD

**Discussion by sberyozkin** [quarkusio/quarkus#42116] [Design Discussions]
## Goal

The goal of this working group is to improve `quarkus-oidc` by:

- implementing important new security features
- continuing with improving the way OIDC requests can be intercepted
- prioritizing on some other open issues such as using JWT bearer client authentication to complete the authorization code flow
- engaging with the community and understanding what may have to be simplified to improve a user experience

### Documentation improvements, simplifying dealing with properties
- [ ]  [#43737](https://github.com/quarkusio/quarkus/issues/43737) Expanded OIDC configuration guide
- [ ] [#21075](https://github.com/quarkusio/quarkus/issues/21075) Create Security Guide Flow Diagram
- [ ] [#45142](https://github.com/quarkusio/quarkus/issues/45142) Minimize duplication in OIDC Bearer and Code flow docs
- [ ] [#47240](https://github.com/quarkusio/quarkus/issues/47240) Improve OIDC configuration validation
- [ ] [#45011](https://github.com/quarkusio/quarkus/issues/45011) Provide Vert.x OIDC to Quarkus OIDC migration guides

### Token security elevation

- [ ] [#46213](https://github.com/quarkusio/quarkus/issues/46213) OAuth2 Step Up Authentication
 
### Completed

- [x] [#33059](https://github.com/quarkusio/quarkus/issues/33059) Create OIDC questionnaire
- [x] [#36250](https://github.com/quarkusio/quarkus/issues/38250) OIDC Dynamic Client Registration
- [x] [#39185](https://github.com/quarkusio/quarkus/issues/39185) Migrate OIDC extensions from ConfigGroup to ConfigMapping
- [x] [#42117](https://github.com/quarkusio/quarkus/issues/42117) Add OIDC response fiters
- [x] [#40013](https://github.com/quarkusio/quarkus/issues/40013) Support JWT bearer authentication to complete the code flow
- [x] [#41890](https://github.com/quarkusio/quarkus/issues/41890) Return OIDC code flow error_code and error_description directly to the user in devmode
- [x] [#4482](https://github.com/quarkusio/quarkus/issues/4482) Support for X509 certificate binding for quarkus-oidc
- [x] [#44958](https://github.com/quarkusio/quarkus/issues/44958) Programmatic OIDC setup
- [x] [#42115](https://github.com/quarkusio/quarkus/issues/42115) Support for DPoP for quarkus-oidc

## End in Sight

This work is done when users respond to the OIDC questionnaire and the simplifications tasks, if identified, are completed.
All other listed tasks must be completed as well.

**Comment by cescoffier** [comment]
Thanks @sberyozkin !

Can you tell us more about the "OIDC questionnaire", it's the first time I hear about it (or, I totally forget about ti, which is totally possible).

**Comment by sberyozkin** [comment]
>Can you tell us more about the "OIDC questionnaire", it's the first time I hear about it (or, I totally forget about ti, which is totally possible).

I want to collect feedback from Quarkus OIDC practitioners, who use it in prod, as well as from users who may have tried it. I'd like to understand what their experience is or has been, what features they use or miss, or what they think will help to push it to the next level.

I'll work on preparing a list of questions at #33059 and then once the questions are ready I'll ask in all the channels, where I can, to support the questionnaire. I'm a bit vague about mechanisms to make it work, but I'll figure it out

**Reply by cescoffier** [reply]
Awesome! Thanks! I may have missed the ping on the issue.
I'm looking forward to seeing the results.

**Comment by sberyozkin** [comment]
@cescoffier @maxandersen, I'd like to open a survey next week, I'd appreciate if we can have  an announcement in the Discussions, as not everyone is on X or quarkus-dev.

**Reply by maxandersen** [reply]
Ok, where is the survey ?

**Reply by sberyozkin** [reply]
Hi @maxandersen, 

Shortened URL:

https://forms.gle/KaqdxMwdqb6W7ejP6

or, in the original form, 
https://docs.google.com/forms/d/e/1FAIpQLScx1paouG2NnZ8-dtoLKlhpW2xmso-5ZRf2K3Ho4i9u_mhgqA/viewform?usp=sf_link

You and Clement have an edit access I think. Let me check with @michalvavrik, Michal, can you access it, it is supposed to be a publicly accessible  survey.

Please announce early next week and I'll follow up with X, quarkus-dev and Zulip users messages.

Also, please forward it to anyone else you'd like to provide the feedback

thanks

**Reply by michalvavrik** [reply]
>  Let me check with @michalvavrik, Michal, can you access it, it is supposed to be a publicly accessible survey.

I read it, I can access it. Hope you get it to relevant people and get answers. Thanks

**Reply by sberyozkin** [reply]
Thanks @michalvavrik, sure, hope we can get some feedback if we keep it open for up to 1 month I guess. 
By the way, Max, Clement, as suggested at #33059, if you'd like to propose more questions, please let me know. Anyone is welcome to propose more questions :-)

**Reply by michalvavrik** [reply]
> Anyone is welcome to propose more questions :-)

I was thinking about it and I can't really find any question. The questions there covers required. It always depends on respondent how he approaches it. My only suggestion is to concentrate on Wildfly to Quarkus migration path. Get questions to customers (not only community) that use EAP with Elytron OIDC, ask them whether they experimented with Quarkus (did POCs) and what would make their migration smoother.

**Reply by sberyozkin** [reply]
@michalvavrik I have a question for users who had a Vert.x OIDC experience, I can copy and paste that one and ask the same question about EAP. I'd like to avoid asking very technical questions, for users to finish the whole survey, with all the time constraints for everyone :-)

**Reply by sberyozkin** [reply]
@michalvavrik Instead of creating one more question for WildFly Elytron, and then possibly for other systems like Spring, I've updated the existing Vertx OIDC only question to ask users if they have had an experience with Vertx OIDC or WildFly Elytron, or Spring, and if someone has some relevant experience, hopefully they will provide some feedback

**Reply by michalvavrik** [reply]
I see it, yes, that looks good.

**Reply by sberyozkin** [reply]
@maxandersen Please create an announcement in the next few days, here are the links again:

Shortened URL:

https://forms.gle/KaqdxMwdqb6W7ejP6

or, in the original form,
https://docs.google.com/forms/d/e/1FAIpQLScx1paouG2NnZ8-dtoLKlhpW2xmso-5ZRf2K3Ho4i9u_mhgqA/viewform?usp=sf_link

The proposed announcement title: `Survey on OpenId Connect support`, I propose to keep it open for 1 month, in the end we can summarize the feedback and let users know.

As far as I'm concerned, I don't plan to add more questions after discussing it with Michal and updating one of the questions, but if you or Clement would prefer to ask for something else, let me know and I'll update the form. Otherwise, let's open the survey when you will have time and I'll follow up with linking to it in other channels, thanks

**Reply by maxandersen** [reply]
first time looking - just a few comments:

- The first question "What is your experience with Quarkus OIDC ?" is missing a "No experience" or "other" option with possiblity to put comment.
- the second question is very demanding, maybe write it: "If you chose `We tried to go to production with it but it did not work out` in the previous question, can you outline what did not work:"

rest of questions are ok.

Where is it you want this announced? wouldn't blog and quarkus design discusion post be enough or?

**Reply by sberyozkin** [reply]
@maxandersen Thanks for the feedback, I updated the second question as you proposed; but I could not add `Other` option to the first dropdown list with an option to comment. I added a slightly expanded variation of your proposed `No experience` though, `We have no experience with it but are planning to test it`, because I guess, if someone has no experience with and no any plans about it, they can't complete the rest of the form anyway.

>Where is it you want this announced? wouldn't blog and quarkus design discusion post be enough or?

IMHO, it should be in `Announcements`, otherwise, the design discussion will disappear from the 1st page probably early next week. I'll also support this announcement with X, quarkus-dev and Zulip messages; I'm not sure about it being a blog post...

**Reply by maxandersen** [reply]
> I guess, if someone has no experience with and no any plans about it, they can't complete the rest of the form anyway.

let them be the decider of that. having the info such user provides can be useful and you can always filter it out.

**Reply by sberyozkin** [reply]
@maxandersen Sure :-), it is only `We have no experience with it` now :-)

**Comment by sberyozkin** [comment]
As far at the plan for the OIDC is concerned, apart from whatever we'll consider to look at following the survey results, elevation of the token security is a major theme right now.

I refer to the X509 token binding and DPoP.
We are literally 1 step away now from supporting the former option for FAPI applications,  while the latter is poised to become the main tool for SPA applications, for ex, with Okta just publishing a blog entry about DPoP.

I'm going to prioritise on these 2 issues.

**Comment by sberyozkin** [comment]
Hi @maxandersen, here is a proposed title and text for the survey announcement:

Title: `Survey on OpenId Connect support`
Text:

```
We have opened a survey on the OpenId Connect (OIDC) support in Quarkus and it will remain open for 1 month.

If you have worked with, tried, or considered using the `quarkus-oidc` extension, please complete [this survey](https://forms.gle/KaqdxMwdqb6W7ejP6) and share with us your thoughts about pros and cons of working with and learning about Quarkus OIDC.

Expected survey completion time is 10-20 minutes. The survey is anonymous, you will not be asked to leave your email or other contact details.

Your feedback will be appreciated. 

Thank you !
```

Please tweak it as you'd like and do the announcement, and as I said, I'll follow up with messages in the other channels, thanks.

And, I'll close #33059 the moment the survey goes live :-)

**Reply by maxandersen** [reply]
live here: https://github.com/quarkusio/quarkus/discussions/43650

**Reply by sberyozkin** [reply]
Thanks @maxandersen

**Comment by sberyozkin** [comment]
@maxandersen @cescoffier I've updated the proposed list of actions for this proposed group. I'm not sure if it becomes a WG, but the plan is as follows: 
* Documentation improvements and a programmatic setup - all these tasks are about making it easy to navigate the properties or minimize the necessity of dealing with them
* Providing options for hardening the token secuity: OIDC MTLS binding is nearly complete, and DPoP - to be tackled in 2025

Michal helped with a few of the completed issues, so we have at least 2 people involved.

**Comment by cescoffier** [comment]
Sorry, missed this message. I think we can move it forward? WDYT?

**Comment by sberyozkin** [comment]
@cescoffier Np, this effort is in progress, so if it can become an actual Working Group effort then it will be good.
I'm looking at the Expanded properties guide issue now, and @michalvavrik would like to look at the StepUp Authentication enhancement in a couple of weeks.

**Comment by sberyozkin** [comment]
@BarDweller FYI as well

**Comment by cescoffier** [comment]
Working group officially kicked off:
https://github.com/orgs/quarkusio/projects/46/views/1

**Comment by sberyozkin** [comment]
@cescoffier If possible, please add https://github.com/quarkusio/quarkus/issues/47240 to this WG list of TODO tasks

**Reply by sberyozkin** [reply]
Sorry, I missed the `Add Item` option there

**Reply by cescoffier** [reply]
No problem! 

You can pick any issue and add it to a working group.
You can also create an _item_ directly on the board, which is not a GitHub issue.

**Comment by sberyozkin** [comment]
I'm closing this group as completed and we all add more targeted OIDC related working groups later.
Thanks @cescoffier for setting it up, and @michalvavrik for completing many of the group's tasks

**Reply by sberyozkin** [reply]
@cescoffier I'm not sure how to correctly mark this group resolved (I'm not sure closing at https://github.com/orgs/quarkusio/projects/46/settings is correct), please let me know or resolve if it requires extra permissions.

**Reply by cescoffier** [reply]
I've completed it. Basically, you create a new "status update" with state _complete_.
